lib/xgu/xgu.h

```diff
@@ -0,0 +1,969 @@
+uint32_t* xgu_set_light_specular_color(uint32_t* p, unsigned int light_index, float r, float g, float b) {
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_SPECULAR_COLOR + light_index*128, 3);
+    p = push_float(p, r);
+    p = push_float(p, g);
+    p = push_float(p, b);
+    return p;
+}
+
+inline
+uint32_t* xgu_set_light_local_range(uint32_t* p, unsigned int light_index, float range) {
+    assert(light_index < XGU_LIGHT_COUNT);
+    return push_command_float(p, NV097_SET_LIGHT_LOCAL_RANGE + light_index*128, range);
+}
+
+inline
+uint32_t* xgu_set_light_infinite_half_vector(uint32_t* p, unsigned int light_index, XguVec3 v) {
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_INFINITE_HALF_VECTOR + light_index*128, 3);
+    return push_floats(p, v.f, 3);
+}
+
+inline
+uint32_t* xgu_set_light_infinite_direction(uint32_t* p, unsigned int light_index, XguVec3 v) {
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_INFINITE_DIRECTION + light_index*128, 3);
+    return push_floats(p, v.f, 3);
+}
+
+inline
+uint32_t* xgu_set_light_spot_falloff(uint32_t* p, unsigned int light_index, XguVec3 v) {
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_SPOT_FALLOFF + light_index*128, 3);
+    return push_floats(p, v.f, 3);
+}
+
+inline
+uint32_t* xgu_set_light_spot_direction(uint32_t* p, unsigned int light_index, XguVec4 v) {
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_SPOT_DIRECTION + light_index*128, 4);
+    return push_floats(p, v.f, 4);
+}
+
+inline
+uint32_t* xgu_set_light_local_position(uint32_t* p, unsigned int light_index, XguVec3 v) {
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_LOCAL_POSITION + light_index*128, 3);
+    return push_floats(p, v.f, 3);
+}
+
+inline
+uint32_t* xgu_set_light_local_attenuation(uint32_t* p, unsigned int light_index, float constant_factor, float linear_factor, float quadratic_factor) {
+    assert(light_index < XGU_LIGHT_COUNT);
+    p = push_command(p, NV097_SET_LIGHT_LOCAL_ATTENUATION + light_index*128, 3);
+    p = push_float(p, constant_factor);
+    p = push_float(p, linear_factor);
+    p = push_float(p, quadratic_factor);
+    return p;
+}
+
+/* ==== Texture stuff ==== */
+
+inline
+uint32_t* xgu_set_texture_offset(uint32_t* p, unsigned int texture_index, const void* offset) {
+    assert(texture_index < XGU_TEXTURE_COUNT);
+    return push_command_parameter(p, NV097_SET_TEXTURE_OFFSET + texture_index*64, (uint32_t)offset);
+}
+
+inline
+uint32_t* xgu_set_texture_format(uint32_t* p, unsigned int texture_index, uint8_t context_dma, bool cubemap_enable,
+                                 XguBorderSrc border_src, uint8_t dimensionality,
+                                 XguTexFormatColor format, uint8_t mipmap_levels,
+                                 uint8_t u_size, uint8_t v_size, uint8_t p_size) {
```

Comment:
We'll probably want to change the order of arguments on this function or merge some fields together. However, we'll have to discuss this then.

Don't change order of arguments as it will be very hard to debug such changes - so any breaking API changes will have to be coordinated (an earlier reordering of parameters already broke most of my XGU projects and took me hours to debug).